### PR TITLE
Map editor

### PR DESCRIPTION
**Pull request by Nemikolh**
## Map editor
- [x] webgl component
- [x] Renderer
  - [x] a huge number of tiles
  - [x] a sprite of any size and only one texture
  - [x] basic API to draw stuff with webgl
  - [x] possibility to apply camera based overlay effects
  - [ ] ~~support object rendering~~
  - [ ] ~~heightmap sort~~
- [ ] ~~options~~
  - [x] show grid
  - [ ] ~~show heightmap~~
- [ ] ~~add a matrix library (or quick craft one for our needs)~~
- [ ] ~~back-end support for chipset~~
- [ ] ~~dialog to upload chipset~~
  - [ ] ~~possibility to split into tiles~~
  - [ ] ~~possibility to select an area and mark it as an object~~
  - [ ] ~~for simple object: define the blocking area~~
- [ ] ~~commit a map~~
- [ ] ~~save a map~~
- [x] create a map
- [ ] ~~collaboration on the map~~
- [x] zoom
- [x] pan
- [ ] ~~create/order layers~~
- [x] edition
  - [x] replace tile of a layer by ones contains by the brush
  - [ ] ~~support brush size greater than one tile~~
  - [x] change the content of the brush
- [x] palette
  - [x] open the palette
  - [x] edit brush
  - [ ] ~~text-search for virtual chipset~~

### CODE REVIEW

File: server/apis/auth.ts [hunk 1]
```diff
@@ -1,20 +1,38 @@
...
+        if (!user) return unauthorized(res, 'Invalid user or password');
+        req.logIn(user, (err) => {
+            if (err) {
+                next(err);
+            } else {
+                res.status(200).json({});
+            }
+        });
+    })(req, res, next);
+}, logRequest('has logged in.'));
```
Comment (Nemikolh) [inline server/apis/auth.ts:28]: Erf, I should probably have put that in another PR. :/

File: editor/rendering/tiles.ts [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+
+            // Vertex buffer
+            let index = 0;
+            let indices_index = 0;
+            let eps = 0.1;
+            for (let i = new_ij[0]; i < new_hw[0]; ++i) {
+                for (let j = new_ij[1]; j < new_hw[1]; ++j) {
+                    let first_index = index / 2;
+
+                    vertices[index++] = this.pos[0] + 2 * this.tile_size * j - eps;
```
Comment (Nemikolh) [inline editor/rendering/tiles.ts]: This `2` here is the hack mentioned in the previous commit message. The rendered size is incorrect without it. I haven't find the reason for the bug. It just doesn't make sense.

**Comment by ghost** [comment]
Looks like you are doing a great job here. I will try to test that as soon as possible!

**Comment by Nemikolh** [comment]
Thanks a lot! :smiley_cat:  There's still plenty to do however.

I haven't documented all commands so here is a summary:
- Clicking on a valid location on the map replace the tile with the one under the mouse cursor.
  If the click is maintained and you move the mouse, it update all the tiles that are being encountered.
- If the map has the "focus" (can be required by clicking on the map at an invalid location), then pressing `P` will open the Palette (== virtual chipset described by @Azamung)
- You can zoom with the wheel and pan with the middle mouse button. There is two zoom behaviour: one for the map and one for the Palette.
- In the layers panel, selecting one of the fake image will change the current layer. All tile update will then be affected by this change.

**Comment by ghost** [comment]
Okay. I plan to test all of that during before the end of the week.

**Comment by ghost** [comment]
Okay so I tried to try, but I ran into an error. Both the server and the client spawned correctly, yet when I use my browser (url: `http://localhost:8001`), I got this error:

```
SyntaxError: in strict mode code, functions may be declared only at top level or immediately within another function index.js:24039:22
DEPRECATION WARNING: 'enqueueTask' is no longer supported and will be removed in next major release. Use addTask/addRepeatingTask/addMicroTask libs.js:58987:1
DEPRECATION WARNING: 'dequeueTask' is no longer supported and will be removed in next major release. Use removeTask/removeRepeatingTask/removeMicroTask libs.js:58987:1
[WDS] Warnings while compiling. libs.js:26356:4
./editor/rules/parser.ts
[26, 13]: unused variable: 'globalsInit'
[27, 13]: unused variable: 'locals'
 libs.js:26358:1
```

Note than when I did the `npm install`, I also got that:

```
npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.8
npm WARN mongodb-core@1.2.31 requires a peer of kerberos@~0.0 but none was installed.                        
```

Probably the fault is mine, but if you could help me…

**Comment by ghost** [comment]
Okay so your commit fixes the issue and now, it works (:

> Clicking on a valid location on the map replace the tile with the one under the mouse cursor. If the click is maintained and you move the mouse, it update all the tiles that are being encountered.

**Yes**, it works.

> If the map has the "focus" (can be required by clicking on the map at an invalid location), then pressing P will open the Palette

**Yes**.

> You can zoom with the wheel and pan with the middle mouse button. There is two zoom behaviour: one for the map and one for the Palette.

Sometimes, the edited map disappears and the only way to get it back is to zoom out. Zoom in after that makes the all thing disappars again.

> In the layers panel, selecting one of the fake image will change the current layer. All tile update will then be affected by this change.

layer

I do not know if I am supposed to get that, but when I clicked on it, nothing _seems_ to happen... but maybe I missed something.

In any case, it is a pretty cool work you have done. Can’t wait to see your future work.

**Comment by Nemikolh** [comment]
This PR fixes #9

**Comment by Nemikolh** [comment]
Yay! Epic fight with my understanding of WebGL!! :smile: 

Last amend fix the annoying seams that were present in certain rendering conditions. This is something that is currently visible on the test instance:

seams

This is now fixed! We now have pixel perfect rendering for the map ! :tada: 

Well... almost.

There is a new issue with the texture rendering: the map orientation is wrong. However this should be relatively easier to fix and will come in another commit. :wink:

**Comment by lthms** [comment]
You are awesome, man. Congrats!

**Comment by Nemikolh** [comment]
Okay, I'm not sure if this was a big win in definitive. Now you can randomly see this instead of the previous bug:

new_bug

new_bug_2

I will ask SO to see if someone can help me with that. I've tried every recommendations I've heard / read but nothing made a difference or fixed the bug (some produced even worst results).

**Comment by lthms** [comment]
If I can help, just tell me.

**Comment by Nemikolh** [comment]
Actually, if you could try on your machine to see if you can also see the bug (to trigger it you might need to zoom and pan a lot before it appears).

Thanks a lot!

**Comment by lthms** [comment]
I will try as soon as I can, probably tomorrow night!

**Comment by Nemikolh** [comment]
For reference, I posted the question here: http://gamedev.stackexchange.com/questions/121051/webgl-pixel-perfect-tilemap-rendering.

Let see if we can get any help! :smiley:

**Comment by lthms** [comment]
Seems like it does not drow much crowds :\

**Comment by lthms** [comment]
So you’ve found your way out? Congrats!

**Comment by Nemikolh** [comment]
Yes! I found an article that contains a pure gpu solution. I originally thought that this would be a slow solution but I should have bench it first. :)
It is remarkably fast! :D

This is not yet finished though. I need to fix the camera and integrate the grid with the new technique. 

Finally I'll do some cleanup. The code has become very messy while I was experimenting.

**Comment by lthms** [comment]
Do you think it’s a good idea to update the online version of dilia?

**Comment by Nemikolh** [comment]
Not yet, I would like to update the sprite rendering first as it does not use the new "way" of computing position relatively to the camera. The two ways of rendering have a slight rendering offset resulting in shaky panning and incorrect overlay tiling.

**Comment by lthms** [comment]
Sounds legit to me. Ping me when you want someone to test your awesome work (:

**Comment by Nemikolh** [comment]
Thinking about it, this PR has been around for so long that it now changes so many part of the code base, it do so much more that was originally planned. Anything new that would done, actually need to be built on top of this PR. I'm thinking that we should merge it now and tackle each task listed here on their own, with small PR per task. That would be helpful to see progress more clearly.

@lethom : What do you think?

**Comment by lthms** [comment]
You got me when you said _their own, with small PR per task_. Do you want me to do some tests here?

**Comment by Nemikolh** [comment]
Yes please! That would be awesome! =)

If you don't find any "real" bug (only features missing), feel free to merge it.
If you find any though, I'll fix it here but without doing anything else (the cleanup will be my next PR).

**Comment by lthms** [comment]
Tonight is the night.
